libraries/tock-cells/src/volatile_cell.rs

```diff
@@ -1,29 +1,41 @@
 //! Implementation of a type for accessing MCU registers.
 
+use core::cell::Cell;
 use core::ptr;
 
 /// `VolatileCell` provides a wrapper around unsafe volatile pointer reads
 /// and writes. This is particularly useful for accessing microcontroller
 /// registers.
-// Source: https://github.com/hackndev/zinc/tree/master/volatile_cell
-#[derive(Copy, Clone, Default)]
-#[repr(C)]
-pub struct VolatileCell<T> {
-    value: T,
+// Originally modified from: https://github.com/hackndev/zinc/tree/master/volatile_cell
+#[derive(Default)]
+#[repr(transparent)]
+pub struct VolatileCell<T: ?Sized> {
+    value: Cell<T>,
+}
+
+impl<T: Copy> Clone for VolatileCell<T> {
+    #[inline]
+    fn clone(&self) -> Self {
+        VolatileCell::new(self.get())
+    }
 }
 
 impl<T> VolatileCell<T> {
     pub const fn new(value: T) -> Self {
-        VolatileCell { value }
+        VolatileCell {
+            value: Cell::new(value),
+        }
     }
 
     #[inline]
-    pub fn get(&self) -> T {
-        unsafe { ptr::read_volatile(&self.value) }
+    pub fn set(&self, value: T) {
```

Comment:
This won't call the destructor for the previous value, but there's no way to do that without a read, defeating the point. Can we make `set` require `T: Copy` to reduce the likelihood of problems due to this, and mention this drop behavior in a doc comment?

Reply:
We can definitely make restrict all of `volatile_cell` to `Copy` types (for MMIO this is almost certainly desirable anyway). But can you explain why this would reduce the likelihood of problems? Is it because it's uncommon for Copy types to have meaningful drop behavior?

(Also, yes, definitely will include this issue in the documentation, great catch!)

Reply:
I've fixed this, but I'm still curious to clarify why requiring `Copy` (which makes sense anyway) helps with the drop issue

Reply:
> Is it because it's uncommon for Copy types to have meaningful drop behavior?

Yes. They're very rare, and when they do exist, their destructors aren't doing any pointer freeing, at least soundly.

Reply:
> > Is it because it's uncommon for Copy types to have meaningful drop behavior?
> 
> Yes. They're very rare, and when they do exist, their destructors aren't doing any pointer freeing, at least soundly.

Not just very rare, nonexistent. No type can implement both `Copy` and `Drop`.